rust/arrow/src/ipc/file/reader.rs

```diff
@@ -0,0 +1,556 @@
+    blocks: Vec<ipc::Block>,
+    /// A counter to keep track of the current block that should be read
+    current_block: usize,
+    /// The total number of blocks, which may contain record batches and other types
+    total_blocks: usize,
+}
+
+impl<R: Read + Seek> Reader<R> {
+    /// Try to create a new reader
+    ///
+    /// Returns errors if the file does not meet the Arrow Format header and footer
+    /// requirements
+    pub fn try_new(reader: R) -> Result<Self> {
+        let mut reader = BufReader::new(reader);
+        // check if header and footer contain correct magic bytes
+        let mut magic_buffer: [u8; 6] = [0; 6];
+        reader.read_exact(&mut magic_buffer)?;
+        if magic_buffer != ARROW_MAGIC {
+            return Err(ArrowError::IoError(
+                "Arrow file does not contain correct header".to_string(),
+            ));
+        }
+        reader.seek(SeekFrom::End(-6))?;
+        reader.read_exact(&mut magic_buffer)?;
+        if magic_buffer != ARROW_MAGIC {
+            return Err(ArrowError::IoError(
+                "Arrow file does not contain correct footer".to_string(),
+            ));
+        }
+        reader.seek(SeekFrom::Start(8))?;
+        // determine metadata length
+        let mut meta_size: [u8; 4] = [0; 4];
+        reader.read_exact(&mut meta_size)?;
+        let meta_len = u32::from_le_bytes(meta_size);
+
+        let mut meta_buffer = vec![0; meta_len as usize];
+        reader.seek(SeekFrom::Start(12))?;
+        reader.read_exact(&mut meta_buffer)?;
+
+        let vecs = &meta_buffer.to_vec();
+        let c_message = ipc::get_root_as_message(vecs);
+        // message header is a Schema, so read it
+        let c_schema: ipc::Schema = c_message.header_as_schema().unwrap();
+        let schema = ipc::convert::fb_to_schema(c_schema);
+
+        // what does the footer contain?
+        let mut footer_size: [u8; 4] = [0; 4];
+        reader.seek(SeekFrom::End(-10))?;
+        reader.read_exact(&mut footer_size)?;
+        let footer_len = u32::from_le_bytes(footer_size);
+
+        // read footer
+        let mut footer_data = vec![0; footer_len as usize];
+        reader.seek(SeekFrom::End(-10 - footer_len as i64))?;
+        reader.read_exact(&mut footer_data)?;
+        let c_footer = ipc::get_root_as_footer(&footer_data[..]);
+
+        let c_blocks = c_footer.recordBatches().unwrap();
+
+        let total_blocks = c_blocks.len();
+
+        Ok(Self {
+            reader,
+            schema: Arc::new(schema),
+            blocks: c_blocks.to_vec(),
+            current_block: 0,
+            total_blocks,
+        })
+    }
+
+    /// Return the number of batches in the file
+    pub fn num_batches(&self) -> usize {
+        self.total_blocks
+    }
+
+    /// Return the schema of the file
+    pub fn schema(&self) -> Arc<Schema> {
+        self.schema.clone()
+    }
+
+    /// Read the next record batch
```

Comment:
Will this remain appropriate as returning `RecordBatch`, given that we can return dictionaries and tensors in the future?

Reply:
Good point. For dictionary it is OK as we use the dictionary to decode record batches. We may need additional methods to read tensor/sparse tensor. We can use the [C++ implementation](https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/reader.h) as a reference.